sparse_strips/vello_common/src/strip.rs

```diff
@@ -38,103 +39,63 @@ pub fn render(
-                        // y values will be 1 if the point is below the current row,
-                        // 0 if the point is above the current row, and between 0-1
-                        // if it is on the same row.
-                        let y0 = rel_y.clamp(0.0, 1.0);
-                        let y1 = (p1.y - y as f32).clamp(0.0, 1.0);
-                        // If != 0, then the line intersects the current row
-                        // in the current tile.
-                        let dy = y0 - y1;
-
-                        // x intersection points in the current tile.
-                        let xx0 = rel_x + (y0 - rel_y) * inv_slope;
-                        let xx1 = rel_x + (y1 - rel_y) * inv_slope;
-                        let xmin0 = xx0.min(xx1);
-                        let xmax = xx0.max(xx1);
-                        // Subtract a small delta to prevent a division by zero below.
-                        let xmin = xmin0.min(1.0) - 1e-6;
-                        // Clip x_max to the right side of the pixel.
-                        let b = xmax.min(1.0);
-                        // Clip x_max to the left side of the pixel.
-                        let c = b.max(0.0);
-                        // Clip x_min to the left side of the pixel.
-                        let d = xmin.max(0.0);
-                        // Calculate the covered area.
-                        // TODO: How is this formula derived?
-                        let mut a = (b + 0.5 * (d * d - c * c) - xmin) / (xmax - xmin);
-                        // a can be NaN if dy == 0 (and thus xmax - xmin = 0, resulting in
-                        // a division by 0 above). This code changes those NaNs to 0.
-                        a = a.abs().max(0.).copysign(a);
-
-                        areas[x as usize][y] += a * dy;
-
-                        // Making this branchless doesn't lead to any performance improvements
-                        // according to my measurements.
-                        if p0.x == 0.0 {
-                            areas[x as usize][y] += (y as f32 - p0.y + 1.0).clamp(0.0, 1.0);
-                        } else if p1.x == 0.0 {
-                            areas[x as usize][y] -= (y as f32 - p1.y + 1.0).clamp(0.0, 1.0);
-                        }
-                    }
-                }
-            }
+    // The strip we're building.
+    let mut strip = Strip {
+        x: prev_tile.x * Tile::WIDTH as i32,
+        y: prev_tile.y * Tile::HEIGHT,
+        col: alpha_buf.len() as u32,
+        winding: 0,
+    };
+
+    for tile in tiles.iter().copied().chain([SENTINEL]) {
+        let line = lines[tile.line_idx as usize];
+        let tile_left_x = tile.x as f32 * Tile::WIDTH as f32;
+        let tile_top_y = tile.y as f32 * Tile::HEIGHT as f32;
+        let p0_x = line.p0.x - tile_left_x;
+        let p0_y = line.p0.y - tile_top_y;
+        let p1_x = line.p1.x - tile_left_x;
+        let p1_y = line.p1.y - tile_top_y;
```

Comment:
With these calculations placed at the front of the loop, strip generation measures at ~1-2% faster for paris-30k on my machine.